python/cuspatial/cuspatial/geometry/geoseries.py

```diff
@@ -0,0 +1,457 @@
+class GpuPolygons(GpuOffset):
+    def __init__(self):
+        self.polys = None
+        self.rings = None
+        self.mpolys = None
+
+    def __repr__(self):
+        result = ""
+        result += "xy:\n" + self.xy.__repr__() + "\n"
+        result += "polys:\n" + self.polys.__repr__() + "\n"
+        result += "rings:\n" + self.rings.__repr__() + "\n"
+        result += "mpolys:\n" + self.mpolys.__repr__() + "\n"
+        return result
+
+
+class cuGeometry:
+    def __init__(self, source, index):
+        self.source = source
+        self.index = index
+
+
+class cuPoint(cuGeometry):
+    def to_shapely(self):
+        item_type = self.source.types[self.index]
+        index = (
+            pd.Series(self.source.types[0 : self.index]) == item_type
+        ).sum()
+        return Point(self.source._points[index].reset_index(drop=True))
+
+
+class cuMultiPoint(cuGeometry):
+    def to_shapely(self):
+        item_type = self.source.types[self.index]
+        item_length = self.source.lengths[self.index]
+        item_start = (
+            pd.Series(self.source.types[0 : self.index]) == item_type
+        ).sum()
+        item_source = self.source._multipoints
+        result = item_source[item_start]
+        return MultiPoint(result.to_array().reshape(item_length, 2))
+
+
+class cuLineString(cuGeometry):
+    def to_shapely(self):
+        ml_index = self.index - 1
+        preceding_line_count = 0
+        preceding_ml_count = 0
+        while ml_index >= 0:
+            if self.source.types[ml_index] == "ml":
+                preceding_ml_count = preceding_ml_count + 1
+            elif (
+                self.source.types[ml_index] == "l" and preceding_ml_count == 0
+            ):
+                preceding_line_count = preceding_line_count + 1
+            ml_index = ml_index - 1
+        preceding_multis = preceding_ml_count
+        if preceding_multis > 0:
+            multi_end = self.source._lines.mlines[preceding_multis * 2 - 1]
+            item_start = multi_end + preceding_line_count
+        else:
+            item_start = preceding_line_count
+        item_length = self.source.lengths[self.index]
+        item_end = item_length + item_start
+        item_source = self.source._lines
+        result = item_source[item_start:item_end]
+        return LineString(
+            result.to_array().reshape(2 * (item_start - item_end), 2)
+        )
+
+
+class cuMultiLineString(cuGeometry):
+    def to_shapely(self):
+        item_type = self.source.types[self.index]
+        index = (
+            pd.Series(self.source.types[0 : self.index]) == item_type
+        ).sum()
+        line_indices = slice(
+            self.source._lines.mlines[index * 2],
+            self.source._lines.mlines[index * 2 + 1],
+        )
+        lines = []
+        for i in range(line_indices.start, line_indices.stop, 1):
+            line = self.source._lines[i].to_array()
+            lines.append(LineString(line.reshape(int(len(line) / 2), 2)))
```

Comment:
nit: Could turn this into a List comprehension instead of initialize then append.

Reply:
done